scalalib/src/JavaModule.scala

```diff
@@ -577,39 +607,44 @@ trait JavaModule
 
   /**
    * Command to print the transitive dependency tree to STDOUT.
-   *
-   * @param inverse Invert the tree representation, so that the root is on the bottom.
-   * @param withCompile Include the compile-time only dependencies (`compileIvyDeps`, provided scope) into the tree.
-   * @param withRuntime Include the runtime dependencies (`runIvyDeps`, runtime scope) into the tree.
-   */
-  def ivyDepsTree(
-      inverse: Boolean = false,
-      withCompile: Boolean = false,
-      withRuntime: Boolean = false
-  ): Command[Unit] =
-    (withCompile, withRuntime) match {
-      case (true, true) =>
-        T.command {
-          printDepsTree(
-            inverse,
-            T.task {
-              transitiveCompileIvyDeps() ++ runIvyDeps()
-            }
-          )
-        }
-      case (true, false) =>
-        T.command {
-          printDepsTree(inverse, transitiveCompileIvyDeps)
-        }
-      case (false, true) =>
-        T.command {
-          printDepsTree(inverse, runIvyDeps)
-        }
-      case _ =>
-        T.command {
-          printDepsTree(inverse, T.task { Agg.empty[Dep] })
-        }
+   */
+  def ivyDepsTree(args: IvyDepsTreeArgs): Command[Unit] = {
+
+    val dependsOnModules = args.whatDependsOn.map(ModuleParser.javaOrScalaModule(_))
+
+    val (invalidModules, validModules) =
+      args.whatDependsOn.map(ModuleParser.javaOrScalaModule(_)).partitionMap(identity)
+
+    if (invalidModules.isEmpty) {
+      (args.withCompile, args.withRuntime) match {
+        case (Flag(true), Flag(true)) =>
+          T.command {
+            printDepsTree(
+              args.inverse.value,
+              T.task {
+                transitiveCompileIvyDeps() ++ runIvyDeps()
+              },
+              validModules
+            )
+          }
+        case (Flag(true), Flag(false)) =>
+          T.command {
+            printDepsTree(args.inverse.value, transitiveCompileIvyDeps, validModules)
+          }
+        case (Flag(false), Flag(true)) =>
+          T.command {
+            printDepsTree(args.inverse.value, runIvyDeps, validModules)
+          }
+        case _ =>
+          T.command {
+            printDepsTree(args.inverse.value, T.task { Agg.empty[Dep] }, validModules)
+          }
+      }
+    } else {
+      val msg = invalidModules.mkString("\n")
+      T.command(println(msg))
```

Comment:
So I'm not 100% sure how to handle this. I think this should actually be a failure, but I wasn't really sure how to fail here. I wanted something like `Result.Failure(msg)`, but that won't work here. How do I fail if it expects a `Comman[Unit]` here?

Reply:
I initial thought was, that something like this should work.

```suggestion
      T.command {
        val msg = invalidModules.mkString("\n")
        Result.Failure(msg)
      }
```

What was the concrete error message? If it's only type inference, maybe a `Result.Failure[Unit](msg)` is enough?

Reply:
ahh adding in the explicit `Unit` helped. I wasn't doing that when I tried and was getting:

```
ambiguous implicit values:
 both value pathReadWrite in trait JsonFormatters of type upickle.default.ReadWriter[os.Path]
 and value regexReadWrite in trait JsonFormatters of type upickle.default.ReadWriter[scala.util.matching.Regex]
 match expected type upickle.default.Writer[T]
```

So what you recommended works. Thanks!

Reply:
Yeah, because it crosses a macro boundary (`T.command`), type inference isn't as helpful as usual.